File: java/core/src/test/java/network/aika/activations/ActivationTest.java
package network.aika.activations;


import network.aika.Document;
import network.aika.bindingsignal.BindingSignal;
import network.aika.neurons.Neuron;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.Map;

import static network.aika.activations.TestBSTypes.A;
import static network.aika.activations.TestBSTypes.B;
import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.mock;


@ExtendWith(MockitoExtension.class)
public class ActivationTest extends AbstractActivationTest {


    @Test
    public void testHasConflictingBindingSignals() {
        Document doc = mock(Document.class);
        Neuron n = mock(Neuron.class);

        BindingSignal bs0 = new BindingSignal(0, doc);
        BindingSignal bs1 = new BindingSignal(1, doc);

        Activation act = new ConjunctiveActivation(
                null,
                null,
                1,
                n,
                doc,
                Map.of(A, bs0)
        );

        assertFalse(
                act.hasConflictingBindingSignals(
                        Map.of(A, bs0)
                )
        );

        assertFalse(
                act.hasConflictingBindingSignals(
                        Map.of(
                                A, bs0,
                                B, bs1
                        )
                )
        );

        assertTrue(
                act.hasConflictingBindingSignals(
                        Map.of(
                                A, bs1,
                                B, bs0
                        )
                )
        );
    }

    @Test
    public void testHasNewBindingSignals() {
        Document doc = mock(Document.class);
        Neuron n = mock(Neuron.class);

        BindingSignal bs0 = new BindingSignal(0, doc);
        BindingSignal bs1 = new BindingSignal(1, doc);

        Activation act = new ConjunctiveActivation(
                null,
                null,
                1,
                n,
                doc,
                Map.of(A, bs0)
        );

        assertTrue(
                act.hasNewBindingSignals(
                        Map.of(
                                A, bs0,
                                B, bs1
                        )
                )
        );

        assertFalse(
                act.hasNewBindingSignals(
                        Map.of(A, bs0)
                )
        );
    }

    @Test
    public void testBranch() {
        BindingSignal bs0 = new BindingSignal(0, doc);
        BindingSignal bs1 = new BindingSignal(1, doc);

        Activation parentAct = new ConjunctiveActivation(
                null,
                null,
                1,
                neuron,
                doc,
                Map.of(A, bs0)
        );

        Activation childAct = parentAct.branch(
                Map.of(B, bs1)
        );

        assertEquals(parentAct, childAct.getParent());
        assertEquals(1, childAct.getBindingSignals().size());
        assertEquals(bs1, childAct.getBindingSignal(B));
    }

    @Test
    public void testCollectLinkingTargets() {

    }

    @Test
    public void testLinkOutgoing() {

    }

    @Test
    public void testPropagate() {

    }
}

File: python-tests/activation_test.py
import unittest
import sys
import os

# Add the project root to Python's module search path
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))

import aika
from abstract_activation_test import AbstractActivationTest, TestBSType, TestUtils

class ActivationTest(AbstractActivationTest):
    
    def test_has_conflicting_binding_signals(self):
        """Test hasConflictingBindingSignals method"""
        # Create binding signals
        bs0 = aika.network.BindingSignal(0, self.doc)
        bs1 = aika.network.BindingSignal(1, self.doc)
        
        # Create activation with binding signal A -> bs0
        binding_signals = {self.BSType_A: bs0}
        activation = aika.network.ConjunctiveActivation(
            self.activation_def,
            None,  # parent
            1,     # id
            self.neuron,
            self.doc,
            binding_signals
        )
        
        # Test 1: Same binding signal should not conflict
        target_signals = {self.BSType_A: bs0}
        self.assertFalse(activation.hasConflictingBindingSignals(target_signals))
        
        # Test 2: Additional binding signal should not conflict
        target_signals = {self.BSType_A: bs0, self.BSType_B: bs1}
        self.assertFalse(activation.hasConflictingBindingSignals(target_signals))
        
        # Test 3: Different binding signal for same type should conflict
        target_signals = {self.BSType_A: bs1, self.BSType_B: bs0}
        self.assertTrue(activation.hasConflictingBindingSignals(target_signals))
    
    def test_has_new_binding_signals(self):
        """Test hasNewBindingSignals method"""
        # Create binding signals
        bs0 = aika.network.BindingSignal(0, self.doc)
        bs1 = aika.network.BindingSignal(1, self.doc)
        
        # Create activation with binding signal A -> bs0
        binding_signals = {self.BSType_A: bs0}
        activation = aika.network.ConjunctiveActivation(
            self.activation_def,
            None,  # parent
            1,     # id
            self.neuron,
            self.doc,
            binding_signals
        )
        
        # Test 1: Additional binding signal should be detected as new
        target_signals = {self.BSType_A: bs0, self.BSType_B: bs1}
        self.assertTrue(activation.hasNewBindingSignals(target_signals))
        
        # Test 2: Same binding signals should not be detected as new
        target_signals = {self.BSType_A: bs0}
        self.assertFalse(activation.hasNewBindingSignals(target_signals))
    
    def test_branch(self):
        """Test branch method"""
        # Create binding signals
        bs0 = aika.network.BindingSignal(0, self.doc)
        bs1 = aika.network.BindingSignal(1, self.doc)
        
        # Create parent activation
        parent_binding_signals = {self.BSType_A: bs0}
        parent_activation = aika.network.ConjunctiveActivation(
            self.activation_def,
            None,  # parent
            1,     # id
            self.neuron,
            self.doc,
            parent_binding_signals
        )
        
        # Create child activation through branching
        child_binding_signals = {self.BSType_B: bs1}
        child_activation = parent_activation.branch(child_binding_signals)
        
        # Verify parent-child relationship
        self.assertEqual(parent_activation, child_activation.getParent())
        
        # Verify child has the new binding signal
        child_signals = child_activation.getBindingSignals()
        self.assertEqual(1, len(child_signals))
        self.assertEqual(bs1, child_activation.getBindingSignal(self.BSType_B))
    
    def test_collect_linking_targets(self):
        """Test collectLinkingTargets method"""
        # This test would require more complex setup with multiple neurons
        # For now, we'll create a placeholder test
        pass
    
    def test_link_outgoing(self):
        """Test linkOutgoing method"""
        # This test would require synapse setup
        # For now, we'll create a placeholder test
        pass
    
    def test_propagate(self):
        """Test propagate method"""
        # This test would require synapse setup
        # For now, we'll create a placeholder test
        pass


if __name__ == '__main__':
    unittest.main()